FEM-2272 upgrade to Exo2.9.1

giladna opened the pull request.

giladna committed:
compile 2.9.0

giladna committed:
use new CustomVideoCodecRenderer Code based from ExoPlayer

giladna committed:
fix adaptive audio bitrates were not added to list

giladna committed:
ignore incorrect video tracks

giladna committed:
fix deprecated methods usage

giladna committed:
remove empty line

giladna committed:
remove - deprecated - done automatically now

giladna committed:
fix deprecated calls

giladna committed:
keep current code with out decoders tests

giladna committed:
upgrade to 2.9.0
support

    private int backBufferDurationMs = DEFAULT_BACK_BUFFER_DURATION_MS;
    private boolean retainBackBufferFromKeyframe = DEFAULT_RETAIN_BACK_BUFFER_FROM_KEYFRAME;

in LoadControl

Support in Language Label in Dash

giladna committed:
add retry API
add assertPlayerIsNotNull(String methodName)  to simplify the null checks

giladna committed:
fix the comment

giladna committed:
cosmetics

giladna committed:
Codacy suggestions

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/DummySurfaceWorkaroundTest.java:
```diff
@@ -117,7 +116,31 @@ private static void workaroundRequired(Context context, boolean b) {
         try {
-            return DefaultDrmSessionManager.newWidevineInstance(fakeDrmCallback, null, mainHandler, null);
+            DefaultDrmSessionManager defaultDrmSessionManager = DefaultDrmSessionManager.newWidevineInstance(fakeDrmCallback, null);
+            if (mainHandler != null) {
```
Remove this section, don't need a listener.

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java:
```diff
@@ -480,15 +478,19 @@ private SelectionOverride retrieveOverrideSelection(int[] uniqueId) {
-
-                        if (audioGroupIndex == groupIndex && audioTrackIndex != TRACK_ADAPTIVE) {
-                            adaptiveTrackIndexesList.add(getIndexFromUniqueId(audioTrack.getUniqueId(), TRACK_INDEX));
+                        //TODO  - Validate this logic with more streams
```
Test with:
"http://cdnapi.kaltura.com/p/2215841/playManifest/entryId/1_w9zx2eti/format/mpegdash/protocol/http/a.mpd"
"http://cdnapi.kaltura.com/p/2215841/playManifest/entryId/1_w9zx2eti/format/applehttp/protocol/http/a.m3u8"

giladna replied:
looks fine with these streams

noamtamim reviewed (changes requested):
Check offline+DRM.

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/Player.java:
```diff
@@ -204,6 +204,13 @@
      */
     void replay();
 
+    /**
```
Remove retry.

giladna replied:
done

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/drm/MediaDrmSession.java:
```diff
@@ -63,7 +67,10 @@ void restoreKeys(byte[] keySetId) {
     FrameworkMediaDrm.KeyRequest getOfflineKeyRequest(byte[] initData, String mimeType) {
         try {
-            return mMediaDrm.getKeyRequest(mSessionId, initData, mimeType, MediaDrm.KEY_TYPE_OFFLINE, null);
+            DrmInitData.SchemeData schemeData = new DrmInitData.SchemeData(MediaSupport.WIDEVINE_UUID, mimeType, initData);
```
Use Collections.singletonList().

giladna replied:
done

giladna committed:
remove retry api for now

giladna committed:
remove unused listener

giladna committed:
add space

giladna committed:
fix warning regarding catch is empty

giladna committed:
remove extra ;

giladna committed:
remove extra ;

giladna committed:
fix warninig

giladna committed:
protect from accessing null referance also for isAdDisplayed

giladna committed:
fix warning for npe

giladna committed:
use singletonList instead of List

giladna committed:
remove TODO

giladna committed:
fix deprecation of get api in DrmInitData

giladna committed:
fix warnings

giladna committed:
remove unused parameter

giladna committed:
add protected explicitly

noamtamim reviewed (changes requested):
Let's upgrade to 2.9.1 before merging, there are fixes that may be important.

giladna committed:
use ExoPlayer 2.9.1

giladna committed:
remove unneeded handler
add PKCodecSupport so if ExoPlayer's isCodecSupported has a bug and result is incorrect -? it will check also codecs logic

giladna committed:
remove unused imports

giladna committed:
simplify getPreferredTrackId method

giladna committed:
instead of firing illegal state exception will fire PKError that contains this exception with new Error Type TRACK_SELECTION_FAILED

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java:
```diff
@@ -617,13 +617,12 @@ private boolean isAdaptive(int rendererIndex, int groupIndex) {
+     * @return IllegalArgumentException when uniqueId is illegal ele return null
      */
-    private int[] validateUniqueId(String uniqueId) throws IllegalArgumentException {
+    public IllegalArgumentException validateUniqueId(String uniqueId) {
```
Returning an exception from a function looks weird. Maybe instead, throw the message like before but be prepared to catch it in ExoPlayerWrapper.
So, no change in TrackSelectionHelper (make it like it was before this commit) and in ExoPlayerWrapper add `try..catch (IllegalArgumentException)` around changeTrack().

giladna replied:
done

giladna committed:
use try catch and upon exception catch it and fire PKError

giladna committed:
need to execute this code block only if languageText == null
before return was called

giladna committed:
fix warnings
